bittide/src/Bittide/Transceiver.hs

```diff
@@ -0,0 +1,260 @@
+        chan clkPath
+        rxn
+        rxp
+
+        freeclk -- gtwiz_reset_clk_freerun_in
+
+        rst_all
+        noReset -- gtwiz_reset_tx_pll_and_datapath_in
+        noReset -- gtwiz_reset_tx_datapath_in
+        noReset -- gtwiz_reset_rx_pll_and_datapath_in
+        rst_rx -- gtwiz_reset_rx_datapath_in
+        gtwiz_userdata_tx_in
+        txctrl2
+        -- gtrefclk -- gtrefclk00_in
+        freeclk -- drpclk_in
+        gtrefclk -- gtrefclk0_in
+
+  (gtwiz_userdata_tx_in,txctrl2) = prbsStimuliGen tx_clk txStimRst
+  rstPrbsChk = resetGlitchFilter (SNat @125000) rx_clk (unsafeFromHighPolarity $ unsafeSynchronizer freeclk rx_clk (rst_prbsChk .||. vio_rst_prbsChk))
+
+  prbsErrors = C.withClockResetEnable rx_clk rstPrbsChk enableGen prbsChecker prbsConf31w64 rx_data
+  anyErrors = fmap (pack . reduceOr) prbsErrors
+  link_up = linkStateTracker rx_clk noReset anyErrors
+
+  rst_all_in = resetGlitchFilter (SNat @125000) freeclk (unsafeFromHighPolarity (rst_all_btn .||. vio_rst_all_btn)) -- $ unsafeSynchronizer freeclk rx_clk sw_n)
+
+  (rst_all, rst_rx, init_done) = gthResetManager freeclk rst_all_in rx_clk (unpack <$> reset_tx_done) (unpack <$> reset_rx_done) link_up
+
+  txStimRst' = resetGlitchFilter (SNat @125000) tx_clk (unsafeFromHighPolarity $ unsafeSynchronizer freeclk tx_clk (rst_txStim .||. vio_rst_txStim))
+  txStimRst = resetSynchronizer tx_clk $ orReset txStimRst' (unsafeFromLowPolarity $ fmap bitCoerce tx_active)
+
+  (  fmap unpack -> vio_rst_all_btn
+   , fmap unpack -> vio_rst_txStim
+   , fmap unpack -> vio_rst_prbsChk
+   ) = unbundle $ vioProbe ("rst_all" :> "rst_rx" :> "init_done" :> Nil) ("rst_all_out" :> "rst_txStim" :> "rst_prbsChk" :> Nil) (0,0,0) freeclk (pack <$> unsafeToHighPolarity rst_all) (pack <$> unsafeToHighPolarity rst_rx) (pack <$> init_done)
+
+orReset :: KnownDomain dom => Reset dom -> Reset dom -> Reset dom
+orReset a b = unsafeFromHighPolarity (unsafeToHighPolarity a .||. unsafeToHighPolarity b)
+
+
+noReset :: KnownDomain dom => Reset dom
+noReset = unsafeFromHighPolarity (pure False)
+
+data LinkSt = Down | Up deriving (Eq,Show,Generic,NFDataX)
+type LinkStCntr = Index 127
+
+linkStateTracker :: (KnownDomain dom, KnownNat w) => Clock dom -> Reset dom -> Signal dom (BitVector w) -> Signal dom Bool
+linkStateTracker clk rst x = fst $ linkStateTracker' clk rst x
+linkStateTracker' :: (KnownDomain dom, KnownNat w) => Clock dom -> Reset dom -> Signal dom (BitVector w) -> (Signal dom Bool, Signal dom LinkStCntr)
+linkStateTracker' clk rst = mooreB clk rst enableGen update genOutput initSt . (reduceOr <$>)
+ where
+  initSt = (Down,0)
+  update :: (LinkSt,LinkStCntr) -> Bit -> (LinkSt,LinkStCntr)
+  update (st,cntr) anyErrs = case (st,anyErrs) of
+    (Down,1) -> (Down,0)
+    (Down,0) | cntr < maxBound -> (Down, cntr+1)
+             | otherwise       -> (Up,cntr)
+    (Up,0) -> (Up, boundedAdd cntr 1)
+    (Up,1) | cntr > 33 -> (Up, cntr-34)
```

Comment:
These numbers are very magical. They should be defined as a single variable and a comment should be added why this value is the way it is. (Probably: _because Xilinx said so_?)